provider/core.rb

```diff
@@ -70,14 +62,28 @@ class FileTemplate
     attr_accessor :async
     attr_accessor :resource_name
 
-    def initialize(options)
-      options.each { |k, v| method("#{k}=").call(v) }
+    class << self
+      # Construct a new FileTemplate based on a resource object
+      def file_for_resource(output_folder, object, version, config, env)
+        file_template = new(output_folder, object.name, object.__product, version, env)
+        file_template.object = object
+        file_template.config = config
+        file_template
+      end
 
-      @env = {
-        pyformat_enabled: options.dig(:env, :pyformat_enabled),
-        goformat_enabled: options.dig(:env, :goformat_enabled),
-        start_time: options.dig(:env, :start_time)
-      }
+      # Construct a new FileTemplate based on a file to be copied into the output directory
+      def file_for_copy(output_folder, target, api, version, env)
+        new(output_folder, target, api, version, env)
+      end
+    end
+
+    def initialize(output_folder, name, product, version, env)
```

Comment:
```suggestion
    def initialize(output_folder, name, product, version_name, env)
```

Looking at the usages, we expect the name of the version instead of the full version object right?

Reply:
I believe so? Didn't know there was a version object as well, SGTM

Reply:
Well, the field on the FileTemplate itself is called `version`, so I'm inclined to leave it as that unless we want to rename the field itself